File: Documentation/5.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/Querying/HowToProjectQueryResults.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Raven.Client.Documents;
using Raven.Client.Documents.Indexes;
using Raven.Client.Documents.Linq;
using Raven.Client.Documents.Queries;
using Raven.Documentation.Samples.Orders;
using Xunit;

namespace Raven.Documentation.Samples.ClientApi.Session.Querying
{
    public class HowToProjectQueryResults
    {
        public async Task Examples()
        {
            using (var store = new DocumentStore())
            {
                using (var session = store.OpenSession())
                {
                    #region projections_1
                    // request Name, City and Country for all entities from 'Companies' collection
                    var results = session
                        .Query<Company>()
                        .Select(x => new
                        {
                            Name = x.Name,
                            City = x.Address.City,
                            Country = x.Address.Country
                        })
                        .ToList();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region projections_1_async
                    // request Name, City and Country for all entities from 'Companies' collection
                    var results = await asyncSession
                        .Query<Company>()
                        .Select(x => new
                        {
                            Name = x.Name,
                            City = x.Address.City,
                            Country = x.Address.Country
                        })
                        .ToListAsync();
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region projections_2
                    var results = session
                        .Query<Order>()
                        .Select(x => new
                        {
                            ShipTo = x.ShipTo,
                            Products = x.Lines.Select(y => y.ProductName),
                        })
                        .ToList();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region projections_2_async
                    var results = await asyncSession
                        .Query<Order>()
                        .Select(x => new
                        {
                            ShipTo = x.ShipTo,
                            Products = x.Lines.Select(y => y.ProductName),
                        })
                        .ToListAsync();
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region projections_3
                    var results = (from e in session.Query<Employee>()
                                   select new
                                   {
                                       FullName = e.FirstName + " " + e.LastName,
                                   }).ToList();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region projections_3_async
                    var results = await (from e in asyncSession.Query<Employee>()
                                         select new
                                         {
                                             FullName = e.FirstName + " " + e.LastName,
                                         }).ToListAsync();
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region projections_4
                    var results = session
                        .Query<Order>()
                        .Select(x => new
                        {
                            Total = x.Lines.Sum(l => l.PricePerUnit * l.Quantity),
                        })
                        .ToList();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region projections_4_async
                    var results = await asyncSession
                        .Query<Order>()
                        .Select(x => new
                        {
                            Total = x.Lines.Sum(l => l.PricePerUnit * l.Quantity),
                        })
                        .ToListAsync();
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region projections_count_in_projection
                    var results = (from o in session.Query<Order>()
                                    let c = RavenQuery.Load<Company>(o.Company)
                        select new
                        {
                            CompanyName = c.Name,
                            ShippedAt = o.ShippedAt,
                            TotalProducts = o.Lines.Count(), //both empty syntax and with a predicate is supported
                            TotalDiscountedProducts = o.Lines.Count(x => x.Discount > 0)
                        }).ToList();
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region projections_5
                    var results = (from o in session.Query<Order>()
                                   let c = RavenQuery.Load<Company>(o.Company)
                                   select new
                                   {
                                       CompanyName = c.Name,
                                       ShippedAt = o.ShippedAt
                                   }).ToList();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region projections_5_async
                    var results = (from o in asyncSession.Query<Order>()
                                   let c = RavenQuery.Load<Company>(o.Company)
                                   select new
                                   {
                                       CompanyName = c.Name,
                                       ShippedAt = o.ShippedAt
                                   }).ToListAsync();
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region projections_6
                    var results = session
                        .Query<Employee>()
                        .Select(e => new
                        {
                            DayOfBirth = e.Birthday.Day,
                            MonthOfBirth = e.Birthday.Month,
                            Age = DateTime.Today.Year - e.Birthday.Year
                        })
                        .ToList();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region projections_6_async
                    var results = await asyncSession
                        .Query<Employee>()
                        .Select(e => new
                        {
                            DayOfBirth = e.Birthday.Day,
                            MonthOfBirth = e.Birthday.Month,
                            Age = DateTime.Today.Year - e.Birthday.Year
                        })
                        .ToListAsync();
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region projections_7
                    var results = from e in session.Query<Employee>()
                                  select new
                                  {
                                      Date = RavenQuery.Raw<DateTime>("new Date(Date.parse(e.Birthday))"),
                                      Name = RavenQuery.Raw(e.FirstName, "substr(0,3)"),
                                  };
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region projections_7_async
                    var results = await (from e in asyncSession.Query<Employee>()
                                         select new
                                         {
                                             Date = RavenQuery.Raw<DateTime>("new Date(Date.parse(e.Birthday))"),
                                             Name = RavenQuery.Raw(e.FirstName, "substr(0,3)"),
                                         }).ToListAsync();
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region projections_8
                    var results = session.Query<Company, Companies_ByContact>()
                        .ProjectInto<ContactDetails>()
                        .ToList();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region projections_8_async
                    var results = await asyncSession.Query<Company, Companies_ByContact>()
                        .ProjectInto<ContactDetails>()
                        .ToListAsync();
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region projections_10
                    // query index 'Products_BySupplierName' 
                    // return documents from collection 'Products' that have a supplier 'Norske Meierier'
                    // project them to 'Products'
                    List<Product> results = session
                        .Query<Products_BySupplierName.Result, Products_BySupplierName>()
                        .Where(x => x.Name == "Norske Meierier")
                        .OfType<Product>()
                        .ToList();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region projections_10_async
                    // query index 'Products_BySupplierName' 
                    // return documents from collection 'Products' that have a supplier 'Norske Meierier'
                    // project them to 'Products'
                    List<Product> results = await asyncSession
                        .Query<Products_BySupplierName.Result, Products_BySupplierName>()
                        .Where(x => x.Name == "Norske Meierier")
                        .OfType<Product>()
                        .ToListAsync();
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region projections_12
                    var results = (from e in session.Query<Employee>()
                                   let format = (Func<Employee, string>)(p => p.FirstName + " " + p.LastName)
                                   select new
                                   {
                                       FullName = format(e)
                                   }).ToList();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region projections_12_async
                    var results = await (from e in asyncSession.Query<Employee>()
                                         let format = (Func<Employee, string>)(p => p.FirstName + " " + p.LastName)
                                         select new
                                         {
                                             FullName = format(e)
                                         }).ToListAsync();
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region projections_13
                    var results = (from e in session.Query<Employee>()
                                   select new
                                   {
                                       Name = e.FirstName,
                                       Metadata = RavenQuery.Metadata(e),
                                   }).ToList();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region projections_13_async
                    var results = await (from e in asyncSession.Query<Employee>()
                                         select new
                                         {
                                             Name = e.FirstName,
                                             Metadata = RavenQuery.Metadata(e),
                                         }).ToListAsync();
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region selectFields
                    var fields = new string[]{
                        "Name",
                        "Phone"
                    };

                    var results = session
                        .Advanced
                        .DocumentQuery<Company, Companies_ByContact>()
                        .SelectFields<ContactDetails>(fields)
                        .ToList();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region selectFields_async
                    var fields = new string[]{
                        "Name",
                        "Phone"
                    };

                    var results = await asyncSession
                        .Advanced
                        .AsyncDocumentQuery<Company, Companies_ByContact>()
                        .SelectFields<ContactDetails>(fields)
                        .ToListAsync();
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region selectFields_2
                    var results = session
                        .Advanced
                        .DocumentQuery<Company, Companies_ByContact>()
                        .SelectFields<ContactDetails>()
                        .ToList();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region selectFields_2_async
                    var results = await asyncSession
                        .Advanced
                        .AsyncDocumentQuery<Company, Companies_ByContact>()
                        .SelectFields<ContactDetails>()
                        .ToListAsync();
                    #endregion
                }
            }
        }

        #region projections_9
        public class Companies_ByContact : AbstractIndexCreationTask<Company>
        {
            public Companies_ByContact()
            {
                Map = companies => companies
                    .Select(x => new
                    {
                        Name = x.Contact.Name,
                        x.Phone
                    });

                StoreAllFields(FieldStorage.Yes); // Name and Phone fields can be retrieved directly from index
            }
        }
        #endregion

        #region projections_9_class
        public class ContactDetails
        {
            public string Name { get; set; }

            public string Phone { get; set; }
        }
        #endregion

        #region projections_11
        public class Products_BySupplierName : AbstractIndexCreationTask<Product>
        {
            public class Result
            {
                public string Name { get; set; }
            }

            public Products_BySupplierName()
            {
                Map =
                    products =>
                        from product in products
                        let supplier = LoadDocument<Supplier>(product.Supplier)
                        select new
                        {
                            Name = supplier.Name
                        };
            }
        }
        #endregion

    }
}

File: Documentation/5.2/Samples/nodejs/ClientApi/Session/Querying/howToProjectQueryResults.js
import * as assert from "assert";
import {DocumentStore, QueryData, AbstractCsharpIndexCreationTask} from "ravendb";

class Product { }
class Companies { }
class Orders { }

const store = new DocumentStore();
const session = store.openSession();

async function examples() {
    {
        //region projections_1
        // request name, city and country for all entities from 'Companies' collection
        const queryData = new QueryData(
            [ "Name", "Address.City", "Address.Country"],
            [ "Name", "City", "Country"]);
        const results = await session
            .query(Companies)
            .selectFields(queryData,Companies)
            .all();
        //endregion
    }

    {
        //region projections_2
        const queryData = new QueryData(
            [ "ShipTo", "Lines[].ProductName" ],
            [ "ShipTo", "Products" ]);

        const results = await session.query(Orders)
            .selectFields(queryData,Orders)
            .all();
        //endregion
    }

    {
        //region projections_3
        const results = await session.advanced
            .rawQuery(
            `from Employees as e select {
                FullName: e.FirstName + " " + e.LastName 
            }`)
            .all();
        //endregion
    }

    {
        //region projections_4
        const results = await session.advanced
            .rawQuery(`from Orders as o select { 
                Total : o.Lines.reduce( 
                    (acc , l) => 
                        acc += l.PricePerUnit * l.Quantity, 0) 
            }`)
            .all();
        //endregion
    }

    {
        //region projections_5
        const results = await session.advanced
            .rawQuery(`from Orders as o 
            load o.Company as c 
            select { 
                CompanyName: c.Name,
                ShippedAt: o.ShippedAt
            }`)
            .all();
        //endregion
    }

    {
        //region projections_6
        const results = await session.advanced
            .rawQuery(
                `from Employees as e 
                select { 
                    DayOfBirth: new Date(Date.parse(e.Birthday)).getDate(), 
                    MonthOfBirth: new Date(Date.parse(e.Birthday)).getMonth() + 1, 
                    Age: new Date().getFullYear() - new Date(Date.parse(e.Birthday)).getFullYear() 
                }`)
            .all();
        //endregion
    }

    {
        //region projections_7
        const results = session.advanced
            .rawQuery(
                `from Employees as e 
                select { 
                    Date : new Date(Date.parse(e.Birthday)), 
                    Name : e.FirstName.substr(0,3) 
                }`)
            .all();
        //endregion
    }

    {
        //region projections_8
        const results = await session.query({ indexName: "Companies/ByContact" })
            .selectFields([ "Name", "Phone" ])
            .all();
        //endregion
    }

    {
        //region projections_10
        // query index 'Products_BySupplierName'
        // return documents from collection 'Products' that have a supplier 'Norske Meierier'
        // project them to instance of Product
        const results = await session.query({ indexName: "Products/BySupplierName" })
            .whereEquals("Name", "Norske Meierier")
            .ofType(Product)
            .all();
        assert.ok(results[0] instanceof Product);
        //endregion
    }

    {
        //region projections_12
        const results = await session.advanced
            .rawQuery(
                `declare function output(e) { 
                    var format = function(p){ return p.FirstName + " " + p.LastName; }; 
                    return { FullName : format(e) }; 
                } 
                from Employees as e select output(e)`)
            .all();
        //endregion
    }

    {
        //region projections_13
        const results = await session.advanced
            .rawQuery(
                `from Employees as e 
                select {
                     Name : e.FirstName, 
                     Metadata : getMetadata(e)
                }`)
            .all();
        //endregion
    }

    //region projections_9_0
    class Companies_ByContact extends AbstractCsharpIndexCreationTask {
        constructor() {
            super();
            this.map = `docs.Company.select(company => new {    
                                Name = company.Contact.Name,    
                                Phone = company.Phone
                            })`;
            this.storeAllFields("Yes"); // name and phone fields can be retrieved directly from index
        }
    }
    //endregion

    //region projections_9_1
    class ContactDetails {
        constructor(name, phone) {
            this.name = name;
            this.phone = phone;
        }
    }
    //endregion

    //region projections_11
    class Products_BySupplierName_Result { }

    class Products_BySupplierName extends AbstractCsharpIndexCreationTask { }
    //endregion

    {
        //region index_10
        class Companies_ByContact  extends AbstractCsharpIndexCreationTask {
            constructor() {
                super();

                this.map = `docs.Company.select(company => new {    
                                Name = company.Contact.Name,    
                                Phone = company.Phone
                            })`;
            }
        }
        //endregion
    }

}
